src/Core/Resolvers/QueryExecutor.cs

```diff
@@ -646,12 +688,36 @@ private async Task<string> GetJsonStringFromDbReader(DbDataReader dbDataReader)
             // 1. https://docs.microsoft.com/en-us/sql/relational-databases/json/format-query-results-as-json-with-for-json-sql-server?view=sql-server-2017#output-of-the-for-json-clause
             // 2. https://stackoverflow.com/questions/54973536/for-json-path-results-in-ssms-truncated-to-2033-characters/54973676
             // 3. https://docs.microsoft.com/en-us/sql/relational-databases/json/use-for-json-output-in-sql-server-and-in-client-apps-sql-server?view=sql-server-2017#use-for-json-output-in-a-c-client-app
-            while (await ReadAsync(dbDataReader))
+
+            if (!_maxResponseSizeLogicEnabled)
+            {
+                while (await ReadAsync(dbDataReader))
+                {
+                    jsonString.Append(dbDataReader.GetString(0));
+                }
+            }
+            else
             {
-                jsonString.Append(dbDataReader.GetString(0));
+                // max buffer size will always be smaller than the 
+                long availableSize = (long)_maxResponseSizeBytes!;
+                while (await ReadAsync(dbDataReader))
+                {
+                    availableSize -= StreamData(dbDataReader, (int)availableSize, jsonString);
+                }
             }
 
             return jsonString.ToString();
         }
+
+        private void ValidateSize(long availableSize, long sizeToBeRead)
+        {
+            if (availableSize - sizeToBeRead < 0)
```

Comment:
why not `sizeToBeRead <= availableSize` ?

Reply:
fixed.